### PR TITLE
New movement!

### PR DESCRIPTION
**Pull request by Bod9001**
### Purpose
to reduce lag to, reduce rubber banding, 
Less spaghetti

generally all-around improvements for movement

### Notes:
StopRandomMidGameCompile_import.cs
I've added this to stop the game compiling while playing ,  note it does pop up a asset importing window,
 it doesn't do anything though You can just move it out of the way and play the game normally  in editor

TODO 
other player movement looks funny ( it gets two messages, set target and force push  it should use set target but it's a pain to work out which is which)

Characters pushing against each other is a bit buggy (even with locker line)
reset doesn't set register tile sometimes?
movement desynchronising  with constant movement ( when it is constantly reset idk if can do this  ) 

rethink space wind pushing player

moving across  matrixs if swap to quickly between multiple , causes appearing in space 
( it's to do with registered tile matrix ID, don't feel like doing it in this PR)

### Changelog:

CL: [New] players can be thrown
CL: [Improvement] Rubber banding should be less severe
CL: [Improvement] server performance should be better

### CODE REVIEW

File: UnityProject/Assets/Scripts/Player/PlayerAction.cs [hunk 1]
```diff
@@ -22,14 +31,55 @@ public Vector2Int Direction()
...
+		else if (direction == Left_Down)
+		{
+			return MovementSynchronisation.PlayerMoveDirection.Down_Left;
+		}
+		else if (direction == Up_Left)
+		{
+			return MovementSynchronisation.PlayerMoveDirection.Up_Right;
+		}
+
+		return MovementSynchronisation.PlayerMoveDirection.Up;
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Player/PlayerAction.cs]: Probably better to use a static dictionary.

Comment (Bod9001) [reply UnityProject/Assets/Scripts/Player/PlayerAction.cs]: ok will do

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/ObjectLayer.cs [hunk 2]
```diff
@@ -98,19 +159,61 @@ public float GetObjectResistanceAt( Vector3Int position, bool isServer )
 	}
 
 
+	public bool CanEnterTileV2(Vector3Int origin, Vector3Int to, bool isServer,List<PushPull> Pushings,  GameObject context = null)
+	{
+		bool PushObjectSet = false;
+		bool CanPushObjects = false;
+
+		//Targeting windoors here
+		foreach (RegisterTile o in GetTileList(isServer).Get(to))
```
Comment (MaxIsJoe) [inline UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/ObjectLayer.cs]: This seems like it's the same foreach loop in 106, Can't you turn this into a function to reduce redundancy?

**Review by MaxIsJoe — CHANGES_REQUESTED** [review]

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/ObjectLayer.cs (hunk 2 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/ObjectLayer.cs]: well is quite small, Can do later

File: UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs [hunk 3]
```diff
@@ -0,0 +1,1631 @@
...
+using Objects;
+using Objects.Construction;
+using UnityEngine;
+using UnityEngine.Events;
+using Util;
+using Random = UnityEngine.Random;
+
+public class UniversalObjectPhysics : NetworkBehaviour, IRightClickable
+{
+	//TODO Push objects being Reset our client
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs]: are all of these TODO still TODO?

File: UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs [hunk 4]
```diff
@@ -0,0 +1,1631 @@
...
+		}
+
+		if (Animating == false && transform.localPosition != NewLocalTarget.Vector3)
+			UpdateManager.Add(CallbackType.UPDATE, AnimationUpdateMe);
+	}
+
+
+	public void StoreTo(ObjectContainer NewParent)
+	{
+		if (NewParent == this) return; //Storing something inside of itself what?
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs]: spawn a blackhole, of course

File: UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs [hunk 5]
```diff
@@ -0,0 +1,1631 @@
...
+		Logger.LogError("registerTile.LocalPosition " + registerTile.LocalPosition.ToString());
+		Logger.LogError("transform.localPosition" + transform.localPosition.ToString());
+
+		Logger.LogError("registerTile.Matrix " + registerTile.Matrix.ToString());
+		Logger.LogError("transform.parent.parent" + transform.parent.parent.ToString());
+	}
+
+
+	[RightClickMethod()]
+	public void ThrowWithSlide()
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs]: are these admin only?

File: UnityProject/Assets/Scripts/Util/RandomUtils.cs [hunk 6]
```diff
@@ -4,6 +4,14 @@
 using UnityEngine;
 using Random = UnityEngine.Random;
 
+public enum SpinMode
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Util/RandomUtils.cs]: feels weird to have this here- It is RandomUtils as in utilities for handling randomization

File: UnityProject/Assets/Scripts/Managers/UpdateManager/ManagedBehaviour.cs [hunk 7]
```diff
@@ -21,7 +21,7 @@ protected virtual void OnDisable()
 	///     If your class uses the Awake function, please use  protected override void Awake() instead.
 	///     Also don't forget to call ManagedNetworkBehaviour.Awake(); first.
 	///     If your class does not use the Awake function, this object will be added to the UpdateManager automatically.
-	///     Do not forget to replace your Update function with public override void UpdateMe()
+	///     Do not forget to replace your Update function with public override void FlyingUpdateMe()
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Managers/UpdateManager/ManagedBehaviour.cs]: `FlyingUpdateMe()`?

Ah I see, global rename or something. There's a few other scripts affected.

File: UnityProject/Assets/Scripts/Messages/Client/RequestMoveMessage.cs [hunk 8]
```diff
@@ -15,7 +15,7 @@ public struct NetMessage : NetworkMessage
 
 		public override void Process(NetMessage msg)
 		{
-			SentByPlayer.Script.PlayerSync.ProcessAction(msg.Action);
+			//SentByPlayer.Script.PlayerSync.ProcessAction(msg.Action);
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Messages/Client/RequestMoveMessage.cs]: So we can remove this script?

File: UnityProject/Assets/Scripts/Messages/Server/InformPullMessage.cs [hunk 9]
```diff
@@ -1,92 +1,2 @@
 using System.Collections.Generic;
 using Mirror;
-
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Messages/Server/InformPullMessage.cs:3]: Can remove this script too I suppose.

File: UnityProject/Assets/Scripts/Objects/Construction/Girder.cs [hunk 10]
```diff
@@ -182,15 +182,15 @@ public void ServerPerformInteraction(HandApply interaction)
 
 		public string Examine(Vector3 worldPos)
 		{
-			return (objectBehaviour.IsPushable ? "Use a wrench to secure to the floor, or a screwdriver to disassemble it."
+			return (objectBehaviour.IsNotPushable! ? "Use a wrench to secure to the floor, or a screwdriver to disassemble it."
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Objects/Construction/Girder.cs]: Exclamation mark at the end? Null forgiving operator... interesting. Though I guess this isn't what we want here.

File: UnityProject/Assets/Scripts/Objects/Gateway/TransportUtility.cs [hunk 11]
```diff
@@ -78,7 +68,7 @@ public static void TransportObjectAndPulled(PushPull pushPullObject, Vector3 tra
 				var previous = node.Previous?.Value;
 
 				//Disconnect pulling to make it not be a problem
-				currentObj.ServerStopPulling();
+				currentObj.PullSet(null, false); //TODO Test without
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Objects/Gateway/TransportUtility.cs:71]: TODO here

File: UnityProject/Assets/Scripts/Objects/ObjectContainer.cs [hunk 12]
```diff
@@ -135,13 +165,13 @@ public void StoreObjects(IEnumerable<GameObject> gameObjects)
...
 		{
-			foreach (var entity in registerTile.Matrix.Get<ObjectBehaviour>(registerTile.LocalPositionServer, true))
+			foreach (var entity in registerTile.Matrix.Get<UniversalObjectPhysics>(registerTile.LocalPositionServer, true))
 			{
 				// Don't add the container to itself...
 				if (entity.gameObject == gameObject) continue;
 
 				// Can't store secured objects (exclude this check on mobs as e.g. magboots set pushable false)
-				if (entity.TryGetComponent<HealthV2.LivingHealthMasterBase>(out _) == false && entity.IsPushable == false) continue;
+				if (entity.IsNotPushable) continue;
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Objects/ObjectContainer.cs:174]: Why remove this check? As far as I'm aware, the comment above still applies.

File: UnityProject/Assets/Scripts/Objects/Other/PortableFlasher.cs [hunk 13]
```diff
@@ -49,7 +49,7 @@ private void WrenchInteraction(PlayerScript wrenchHolder)
 		{
 			isWrenched = !isWrenched;
 			//We inverse this to get the opposite of the wrench, so if its not wrenched; isPushable is true and vice versa
-			objectBehaviour.ServerSetPushable(!isWrenched);
+			objectBehaviour.SetIsNotPushable(!isWrenched);
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Objects/Other/PortableFlasher.cs]:
```suggestion
			objectBehaviour.SetIsNotPushable(isWrenched);
```

File: UnityProject/Assets/Scripts/Player/PlayerScript.cs [hunk 14]
```diff
@@ -380,7 +381,7 @@ public void SyncPlayerName(string oldValue, string value)
 		RefreshVisibleName();
 	}
 
-	public bool IsHidden => !PlayerSync.ClientState.Active;
+	public bool IsHidden => false; //TODO
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Player/PlayerScript.cs]: TODO - does anything use this?

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/RegisterDoor.cs [hunk 15]
```diff
@@ -116,7 +115,7 @@ public override bool IsPassableFromInside(Vector3Int leavingTo, bool isServer, G
 				return !direction.y.Equals(v.y) || !direction.x.Equals(v.x);
 			}
 
-			return !isClosed;
+			return true; //Should be able to walk out of closed doors
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/RegisterDoor.cs:118]: Does this affect one-sided doors?

As an aside, I wanted to take this passable stuff out of RegisterDoor and use DirectionalPassable instead.

File: UnityProject/Assets/Scripts/Player/Movement/PlayerMove.cs [hunk 16]
```diff
@@ -21,7 +21,7 @@ namespace Player.Movement
 	///     handles interaction with objects that can
 	///     be walked into it.
 	/// </summary>
-	public class PlayerMove : NetworkBehaviour, IRightClickable, IServerSpawn, IActionGUI, ICheckedInteractable<ContextMenuApply>, RegisterPlayer.IControlPlayerState
+	public class OLDPlayerMove : NetworkBehaviour, IRightClickable, IServerSpawn, IActionGUI, ICheckedInteractable<ContextMenuApply>, RegisterPlayer.IControlPlayerState
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Player/Movement/PlayerMove.cs]: Anything still use this? Can it be deleted?

File: UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs (hunk 3 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs]: some of them aren't

File: UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs (hunk 4 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs]: hehe

File: UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs (hunk 5 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs]: there meant to be debug only , Easy enough to clean up

File: UnityProject/Assets/Scripts/Managers/UpdateManager/ManagedBehaviour.cs (hunk 7 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Managers/UpdateManager/ManagedBehaviour.cs]: yeah, when Ryder doesn't do a good job

File: UnityProject/Assets/Scripts/Messages/Client/RequestMoveMessage.cs (hunk 8 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Messages/Client/RequestMoveMessage.cs]: should be able to yeah

File: UnityProject/Assets/Scripts/Objects/Construction/Girder.cs (hunk 10 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Objects/Construction/Girder.cs]: == false right?

File: UnityProject/Assets/Scripts/Objects/ObjectContainer.cs (hunk 12 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Objects/ObjectContainer.cs:174]: I removed the player specific storage, and is now just one storage

File: UnityProject/Assets/Scripts/Objects/Other/PortableFlasher.cs (hunk 13 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Objects/Other/PortableFlasher.cs]: isn't the point that it changes each time you interact with it?

File: UnityProject/Assets/Scripts/Player/PlayerScript.cs (hunk 14 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Player/PlayerScript.cs]: oh yeah, need to do that

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/RegisterDoor.cs (hunk 15 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/RegisterDoor.cs:118]: I think it's a separate script for directional doors

Comment (Bod9001) [reply UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/RegisterDoor.cs:118]: return CheckViaDirectional(tryGetDir, direction); 
yeah, is done separately

File: UnityProject/Assets/Scripts/Player/Movement/PlayerMove.cs (hunk 16 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Player/Movement/PlayerMove.cs]: let's see what explodes

File: UnityProject/Assets/Scripts/Objects/Construction/Girder.cs (hunk 10 above)
Comment (corp-0) [reply UnityProject/Assets/Scripts/Objects/Construction/Girder.cs]: in plain English, what do you want to return there?

File: UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs (hunk 3 above)
Comment (corp-0) [reply UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs]: it is probably a better idea to move them into issues (each one), after we merged this

File: UnityProject/Assets/Scripts/Objects/Construction/Girder.cs (hunk 10 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Objects/Construction/Girder.cs]: it should be the opposite of objectBehaviour.IsPushable  would be

Comment (Bod9001) [reply UnityProject/Assets/Scripts/Objects/Construction/Girder.cs]: because it's IsNotPushable

File: UnityProject/Assets/Scripts/Util/RandomUtils.cs (hunk 6 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Util/RandomUtils.cs]: where would be the best place to put it?

**Comment by Bod9001** [comment]
just have to think about object location stuff with inventories

File: UnityProject/Assets/Scripts/Util/RandomUtils.cs (hunk 6 above)
Comment (corp-0) [reply UnityProject/Assets/Scripts/Util/RandomUtils.cs]: somewhere closer to where it is used, which I assume is the UniversalPhysics thingy. Maybe create a new file for all enums it uses? Like we do for health.

Comment (Bod9001) [reply UnityProject/Assets/Scripts/Util/RandomUtils.cs]: could do

File: UnityProject/Assets/Scripts/Objects/Construction/Girder.cs (hunk 10 above)
Comment (NoooneyDude) [reply UnityProject/Assets/Scripts/Objects/Construction/Girder.cs]: Yep `== false` sounds good

File: UnityProject/Assets/Scripts/Objects/ObjectContainer.cs (hunk 12 above)
Comment (NoooneyDude) [reply UnityProject/Assets/Scripts/Objects/ObjectContainer.cs:174]: Hmm I'm not sure what you mean. So the original issue was that players with magboots enabled (which set the player as immovable) couldn't enter lockers - they'd stay outside when the locker closed.

Tbh the magboots shouldn't be setting the player as immovable but be doing something else.

File: UnityProject/Assets/Scripts/Objects/Other/PortableFlasher.cs (hunk 13 above)
Comment (NoooneyDude) [reply UnityProject/Assets/Scripts/Objects/Other/PortableFlasher.cs]: So the issue is that later on the examine message is wrong, it is now inverted.

File: UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs [hunk 17]
```diff
@@ -0,0 +1,1610 @@
...
+	/// Maximum speed player can reach by throwing stuff in space
+	/// </summary>
+	public const float MAX_SPEED = 25;
+
+	public const int HIGH_SPEED_COLLISION_THRESHOLD = 13;
+
+
+	//public const float DEFAULT_Friction = 9999999999999f;
+	public const float DEFAULT_Friction = 15f;
+	public const float DEFAULT_SLIDE_FRICTION = 9f;
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs:41]: Should these be serialised so different objects have different friction or are these some constants needed for some math

File: UnityProject/Assets/Scripts/Messages/Server/PlayerMoveMessage.cs [hunk 18]
```diff
@@ -32,22 +32,22 @@ public override void Process(NetMessage msg)
...
+			// 		playerSync.ClearQueueClient();
+			// 		playerSync.RollbackPrediction();
+			// 	}
+			// 	if (msg.State.MoveNumber == 0 ) {
+			// 		playerSync.ClearQueueClient();
+			// 		playerSync.RollbackPrediction();
+			// 	}
+			//
+			// 	ControlTabs.CheckTabClose();
+			// }
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Messages/Server/PlayerMoveMessage.cs]: Can delete this script too?

File: UnityProject/Assets/Scripts/Objects/Construction/APCFrame.cs [hunk 19]
```diff
@@ -302,7 +302,7 @@ private void WrenchedStateInteraction(HandApply interaction)
 
 				econs.Pool();
 
-				GameObject WallMount = Spawn.ServerPrefab(APCObject, gameObject.GetComponent<CustomNetTransform>().ServerPosition, interaction.Performer.transform.parent, spawnItems: false).GameObject;
+				GameObject WallMount = Spawn.ServerPrefab(APCObject, gameObject.GetComponent<UniversalObjectPhysics>().transform.position, interaction.Performer.transform.parent, spawnItems: false).GameObject;
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Objects/Construction/APCFrame.cs]: Don't need to get the component if we're just using it for `transform.position`.

File: UnityProject/Assets/Scripts/Managers/ComponentManager.cs [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+			{
+				Logger.LogError($"Unable to find UniversalObjectPhysics on {gameObject.name}");
+				return null;
+			}
+
+			ObjectToPhysics[gameObject] = UOP;
+		}
+
+		return ObjectToPhysics[gameObject];
+	}
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Managers/ComponentManager.cs]: So as this could be used a lot we can probably do some little optimisations:

- use `ObjectToPhysics.TryGetValue()` for line `21` so we only do a dictionary lookup once (don't need a lookup on line `33`) (and need to return UOP after line `30` ofc).
- In similar fashion as `TryGetValue()` let's change this method so that it returns a boolean for if found or not, and uses an `out` parameter for the component (like the `TryGetComponent()`s), so that usages of this function don't need to check for null again (Unity makes it expensive).
Edit: actually you return an explicit null, not familiar enough if that changes things - probably not? I think semantically it is more consistent with other `Try` methods to do it as mentioned.

File: UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs (hunk 4 above)
Comment (NoooneyDude) [reply UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs]: You should totally do this 😂 give some developer a laugh if they make a mistake.
And a serverwide message "Something has defeated physics! An ominous feeling washes over you..." so the reason why is easy to find

File: UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs (hunk 17 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs:41]: currently is constant, could be done per object basis,  could even be determined by tile to but Would cause issues with throwing Landing perfectly underneath mouse

File: UnityProject/Assets/Scripts/Objects/Other/PortableFlasher.cs (hunk 13 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Objects/Other/PortableFlasher.cs]: done

File: UnityProject/Assets/Scripts/Objects/ObjectContainer.cs (hunk 12 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Objects/ObjectContainer.cs:174]: yeah, I think going for that approach is good and I can do it as part of rethinking space wind, I'll do that in a separate PR though since it's just a generic push currently

**Review by NoooneyDude — APPROVED** [review]
Codacy has a few good suggestions but otherwise I'm happy 👍

**Review by corp-0 — APPROVED** [review]
